### DISCUSSION TITLE
Proxy session

### DISCUSSION THREAD

**Discussion by DarksideVT** [CrowdStrike/falconpy#980] [Q&A]
How do I get a OAuth2 token (or use any falconpy object) if I have to go through a proxy using Kerberos? I can use requests.Session() and mount the proxy auth but, I'm unable to pass the session to oauth2/token. I could use a post to oauth2/token and return the token but if I try to create a Hosts object with `falconpy.Hosts(creds={"access_token": token}, proxy=proxies, session=session)` it's not getting through the proxy.

**Comment by jshcodes** [comment]
In the example above we're providing `access_token` as a key within the `creds` dictionary.  This should not work, as `access_token` is a separate keyword.  If we change this to be:

```python
falconpy.Hosts(access_token=token, proxy=proxies, session=session)
```

do we get the expected result?

**Reply by DarksideVT** [reply]
The token may be correct in your example but is `session` a keyword that can be passed to the Hosts object?  I ask because the proxy is not accepting the authentication that I used when I got the token initially with `Session.post`

**Reply by jshcodes** [reply]
I just noticed that!  No session is not an available keyword when creating an instance of a Service Class.

Can we have more detail or a programmatic example?  What are you seeing when you try and authenticate via the proxy (without using requests directly)?

**Reply by DarksideVT** [reply]
I'm getting a `Proxy Authentication Required` Error from the response.

**Reply by jshcodes** [reply]
Oh.  The proxy is slapping down the request as _*it*_ needs authentication. 

FalconPy supports providing custom headers to the class on creation, but that might not help here... 🤔 Is the proxy using something like a session cookie to track authorization? Any examples of authenticating via Python laying about?   The solution may involve customizing a class that uses the Service Class you want as a base.

**Reply by DarksideVT** [reply]
Here is how I usually auth a session
Proxy.py
```python
class ProxyKerberose(requests.adapters.HTTPAdapter):
    def send(self, request, stream=False, timeout=None, verify=true, cert=None, proxies=None):
        for delay in [0.1.2.4.-1]:
            ret = super().send(request, stream,timeout,verify,cert,proxies)
            return ret
    def proxy_headers(self,proxy):
        author = requests_kerberose.HTTPKerberoseAuth()
        auth_token = auth_generate_request_header(None, urllib.parse.urlparse(proxy).hostname, is_preenptive=True)
        headers = {'Proxy-Authorization': auth_token}
        return headers
```
main.py
```python
import requests
import Proxy
proxies = {'https': 'http://<proxy URL>', 'http': 'http://<proxy URL>'}
session = requests.Session()
session.proxies = proxies
session.mount('https://', Proxy.ProxyKerberose())
response = session.post('https://<endpoint>', verify=False, data={'example': 'data'}
```

**Reply by jshcodes** [reply]
This feels solvable, but I don't have an environment setup to troubleshoot the specific scenario.  Leaving this open for now while I investigate.

**Reply by DarksideVT** [reply]
I have a way around it with a header. You mentioned a way to insert another header. What's the process?
The auth header looks like this:
`{'Proxy-Authorization': Negotiate <secret>`

**Reply by jshcodes** [reply]
If you're using the Service Class you can provide additional headers using the [`ext_headers`](https://www.falconpy.io/Usage/Environment-Configuration.html#extended-headers) keyword.

So something along the lines of ...
```python
from falconpy import Hosts

extra_header = {"X-SOME-SORT-OF-HEADER": "Some sort of value"}
hosts = Hosts(client_id=CLIENT_ID, client_secret=CLIENT_SECRET, proxy=proxies, ext_headers=extra_header)
```

**Reply by DarksideVT** [reply]
I was afraid of that. I tried that and thought it was wrong because I still get an Auth error from the proxy.

**Reply by jshcodes** [reply]
Is there a way for us to review the request the proxy is receiving?  We can confirm that there isn't an issue with the `ext_headers` keyword.

Do you have a link to the auth header requirements?

**Reply by jshcodes** [reply]
We can also try using the Uber Class, if it's just an issue with `ext_headers`, that _might_ work. (Depends on when the auth event happens. This may need to also be looked at.)

```python
from falconpy import APIHarness

extra_header = {"X-SOME-SORT-OF-HEADER": "Some sort of value"}

uber = APIHarness(client_id=CLIENT_ID, client_secret=CLIENT_SECRET, proxy=proxies)

result = uber.command("QueryDevicesByFilterScroll", headers=extra_header)
```

**Comment by DarksideVT** [comment]
> Is there a way for us to review the request the proxy is receiving? We can confirm that there isn't an issue with the `ext_headers` keyword.
> 
> Do you have a link to the auth header requirements?

Unfortunately, there is no easy way to do this, and it doesn't seem like using the Uber class changes anything.

**Comment by DarksideVT** [comment]
Is there a way to use a kerberos ticket file? That's how other people are getting through the proxy.

**Reply by jshcodes** [reply]
Reading some of the [discussion around this](https://github.com/requests/requests-kerberos/issues/148), it looks to be potentially doable (with some changes and testing).  This is definitely something we should investigate after `1.3.0` merges to main so we're working in the newer codebase.

I'm going to spawn a ticket to look into implementing this.  Stay tuned.

**Reply by DarksideVT** [reply]
Did this get merged?

**Reply by DarksideVT** [reply]
I ask because I added an _proxy.py file and modified the _utils._functions.py locally to get it working on my Kerberos proxy. I'd like to help out with this issue if I can.

**Reply by jshcodes** [reply]
`1.3.0` has merged!  Package release is tomorrow. 🎉  

Submit your contribution to the `dev` branch. This is definitely useful functionality for the SDK to support.  😄 🙇
